### PR TITLE
Worker trait part 4

### PR DESCRIPTION
**Pull request by thibault-martinez**

### CODE REVIEW

File: bee-common-ext/src/node.rs [hunk 1]
```diff
@@ -11,52 +11,79 @@
...
 
 use std::{
     any::TypeId,
     collections::{HashMap, HashSet},
-    sync::Arc,
+    pin::Pin,
 };
 
-pub trait Node: Send + Sync + 'static {
+pub trait Node: Default + Send + Sync + 'static {
```
Comment (zesterer) [inline bee-common-ext/src/node.rs:26]: Good idea

File: bee-common-ext/src/node.rs [hunk 2]
```diff
@@ -11,52 +11,79 @@
...
     where
         Self: Sized,
         W: Worker<Self>,
         G: FnOnce(oneshot::Receiver<()>) -> F,
         F: Future<Output = ()> + Send + Sync + 'static;
+    fn worker<W>(&self) -> Option<&W>
+    where
+        Self: Sized,
+        W: Worker<Self> + Send + Sync;
+    fn set_workers(&mut self, workers: Map<dyn Any + Send + Sync>);
```
Comment (zesterer) [inline bee-common-ext/src/node.rs:38]: Ngl, I wonder whether `register_resource` would be better. Shall I make this change along with my changes?

File: bee-ledger/src/whiteflag/mod.rs [hunk 3]
```diff
@@ -33,41 +35,22 @@ pub fn init(
     index: u32,
     state: LedgerState,
     coo_config: ProtocolCoordinatorConfig,
-    bee_node: Arc<BeeNode>,
+    node_builder: NodeBuilder<BeeNode>,
     bus: Arc<Bus<'static>>,
-    shutdown: &mut Shutdown,
-) -> mpsc::UnboundedSender<LedgerWorkerEvent> {
```
Comment (zesterer) [inline bee-ledger/src/whiteflag/mod.rs:39]: I'm a bit weirded out by this function. Is this a permanent thing?

**Review by zesterer — APPROVED** [review]

File: bee-common-ext/src/node.rs (hunk 2 above)
Comment (thibault-martinez) [reply bee-common-ext/src/node.rs:38]: Yeah let's do as part of your changes

File: bee-ledger/src/whiteflag/mod.rs (hunk 3 above)
Comment (thibault-martinez) [reply bee-ledger/src/whiteflag/mod.rs:39]: The reasoning is that I would like not to publicly expose all of the workers all the time so I prefer to go through an init function. Does that make sense ?

Comment (zesterer) [reply bee-ledger/src/whiteflag/mod.rs:39]: That makes sense. I wonder whether some sort of `with_default_workers` method on `NodeBuilder` would be useful here?

Comment (thibault-martinez) [reply bee-ledger/src/whiteflag/mod.rs:39]: No sure to understand what it would do

Comment (zesterer) [reply bee-ledger/src/whiteflag/mod.rs:39]: Have the `with_default_workers` initiate these workers instead. It's not important at this stage though, feel free to ignore.

**Review by zesterer — APPROVED** [review]

**Review by zesterer — APPROVED** [review]

**Pull request merged**